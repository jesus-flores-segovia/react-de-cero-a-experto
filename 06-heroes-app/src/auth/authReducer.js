import { types } from "../types/types";

export const authReducer = (state = {}, action) => {

    switch (action.type) {
        case types.login:
            return {
                ...state,
                name: action.payload,
                logged: true
            }
        case types.logout:
            return {
                ...state,
                name: "",
                logged: false
            }
    
        default:
            return state;
    }
}