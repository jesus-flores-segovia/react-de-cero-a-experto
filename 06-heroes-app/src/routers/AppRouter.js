import React from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route
} from "react-router-dom";
import { DcScreen } from "../components/dc/DcScreen";
import { LoginScreen } from "../components/login/LoginScreen";
import { MarvelScreen } from "../components/marvel/MarvelScreen";
import { Navbar } from "../components/ui/NavBar";

export const AppRouter = () => {
    return (
        <Router>
          <div>
            <Navbar/>
                <Switch>
                    <Route exact path="/login" component={LoginScreen}/>
                    <Route exact path="/marvel" component={MarvelScreen}/>
                    <Route exact path="/dc" component={DcScreen}/>
                </Switch>
          </div>
        </Router>
      );
    }
