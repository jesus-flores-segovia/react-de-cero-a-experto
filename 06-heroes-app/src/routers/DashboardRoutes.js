import React from 'react';
import {
    Switch,
    Route,
    Redirect
  } from "react-router-dom";
import { HeroScreen } from '../components/heroes/HeroScreen';
import { DcScreen } from "../components/dc/DcScreen";
import { MarvelScreen } from "../components/marvel/MarvelScreen";
import { Navbar } from '../components/ui/NavBar';

export const DashboardRoutes = () => {
    return (
        <>
            <Navbar/>
            <div className="container mt-5">
                <Switch>
                    <Route exact path="/marvel" component={MarvelScreen}/>
                    <Route exact path="/hero/:heroId" component={HeroScreen}/>
                    <Route exact path="/dc" component={DcScreen}/>

                    <Redirect to="/marvel"/>
                </Switch>
            </div>   
        </>
    )
}
