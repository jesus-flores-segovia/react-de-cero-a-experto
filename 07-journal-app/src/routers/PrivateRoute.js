import React from 'react'
import { Navigate } from 'react-router-dom'

export const PrivateRoute = ({
    children,
    isLoggedIn
}) => {
    return !isLoggedIn
        ? <Navigate to="/auth"/>
        : children 
}