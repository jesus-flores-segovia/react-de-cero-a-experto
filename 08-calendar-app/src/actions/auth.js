import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";
import { calendarLogoutCleanEvents } from "./calendar";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithoutToken(
        "auth",
        { email, password },
        "POST"
      );
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      } else {
        if (body.errors) {
          const { email, password } = body.errors;
          const message = email ? email.msg : password.msg;
          Swal.fire("Cannot login", message, "error");
          return;
        }

        Swal.fire("Cannot login", body.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", "An error has occurred", "error");
    }
  };
};

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithoutToken(
        "auth/register",
        { name, email, password },
        "POST"
      );
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      } else {
        if (body.errors) {
          const { name, email, password } = body.errors;
          const message = name ? name.msg : email ? email.msg : password.msg;
          Swal.fire("Cannot register", message, "error");
          return;
        }

        Swal.fire("Cannot register", body.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", "An error has occurred", "error");
    }
  };
};

export const startCheckingLoginState = () => {
  return async (dispatch) => {
    try {
      dispatch(checkLoginState({ isCheckingLoginState: true }));
      const response = await fetchWithToken("auth/newToken");
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      }
      dispatch(checkLoginState({ isCheckingLoginState: false }));
    } catch (error) {
      dispatch(checkLoginState({ isCheckingLoginState: false }));
    }
  };
};

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(calendarLogoutCleanEvents());
    dispatch(logout());
  };
};

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

const checkLoginState = (isCheckingLoginState) => ({
  type: types.authCheckLoginState,
  payload: isCheckingLoginState,
});

const logout = () => ({
  type: types.authLogout,
});
