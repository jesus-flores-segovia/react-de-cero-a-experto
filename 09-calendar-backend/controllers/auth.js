const { response } = require("express");
const { validationResult } = require("express-validator");

const register = (req, res = response) => {
  const { name, email, password } = req.body;
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    return res.status(400).send({ ok: false, errors: errors.mapped() });
  }

  res.status(201).send({ ok: true, msg: "register", name, email, password });
};

const login = (req, res = response) => {
  const { email, password } = req.body;
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    return res.status(400).send({ ok: false, errors: errors.mapped() });
  }

  res.status(201).send({ ok: true, msg: "login", email, password });
};

const newToken = (req, res = response) => {
  res.send({ ok: true, msg: "new token" });
};

module.exports = {
  register,
  login,
  newToken,
};
