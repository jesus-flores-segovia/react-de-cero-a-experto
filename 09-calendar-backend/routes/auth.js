/* 
    Routing for the users auth
    url: host + /api/auth
*/

const express = require("express");
const { check } = require("express-validator");
const router = express.Router();

const { register, login, newToken } = require("../controllers/auth");

router.post(
  "/register",
  [
    check("name", "Name is mandatory").not().isEmpty(),
    check("email", "Email is mandatory").isEmail(),
    check("password", "Password must be at least of 6 characters").isLength({
      min: 6,
    }),
  ],
  register
);

router.post(
  "/",
  [
    check("email", "Email is mandatory").isEmail(),
    check("password", "Password must be at least of 6 characters").isLength({
      min: 6,
    }),
  ],
  login
);

router.get("/newToken", newToken);

module.exports = router;
