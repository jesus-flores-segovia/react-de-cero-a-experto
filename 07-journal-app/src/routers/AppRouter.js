import { onAuthStateChanged } from 'firebase/auth'
import React, { useEffect, useState } from 'react'
import { useDispatch } from 'react-redux'
import { BrowserRouter, Route, Routes, Navigate } from 'react-router-dom'
import { login } from '../actions/auth'
import { startLoadingNotes } from '../actions/notes'

import { JournalScreen } from '../components/journal/JournalScreen'
import { auth } from '../firebase/firebaseConfig'
import { AuthRouter } from './AuthRouter'
import { PrivateRoute } from './PrivateRoute'
import { PublicRoute } from './PublicRoute'

export const AppRouter = () => {

    const dispatch = useDispatch();

    const [checking, setChecking] = useState(true);
    const [isLoggedIn, setIsLoggedIn] = useState(false)

    useEffect(() => {
        onAuthStateChanged(auth, (user) => {
            if(user?.uid) {
                dispatch(login(user.uid, user.displayName));
                setIsLoggedIn(true);

                dispatch(startLoadingNotes());
            } else{
                setIsLoggedIn(false);
            }

            setChecking(false);
        })
    }, [dispatch, setChecking, setIsLoggedIn])

    if(checking) {
        return (
            <h1>Wait...</h1>
        )
    }

    return (
        <BrowserRouter>
            <Routes>
                <Route path="/" 
                    element={
                        <PrivateRoute isLoggedIn={isLoggedIn}>
                            <JournalScreen/>
                        </PrivateRoute>
                        }/>
                <Route path="auth/*" 
                    element={
                        <PublicRoute isLoggedIn={isLoggedIn}>
                            <AuthRouter/>
                        </PublicRoute>
                        }/>
                <Route 
                    path="*"
                    element={<Navigate to="/auth/*" replace={true} />}
                />
            </Routes>
        </BrowserRouter> 
    )
}
