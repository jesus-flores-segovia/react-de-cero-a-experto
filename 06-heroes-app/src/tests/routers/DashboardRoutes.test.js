import { mount } from "enzyme"
import { MemoryRouter } from "react-router"
import { DashboardRoutes } from "../../routers/DashboardRoutes"
import { AuthContext } from "../../auth/AuthContext"

describe("Tests inside file 'DashboardRoutes.test.js'", () => {

    test("DashboardRoutes must be shown correctly", () => {

        const contextValue = {
            dispatch: jest.fn(),
            user: {
                logged: true,
                name:"Jesús"
            }
        }

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter>
                    <DashboardRoutes/>
                </MemoryRouter>
            </AuthContext.Provider>  
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".text-info").text().trim()).toBe("You're logged as Jesús");
    })
})