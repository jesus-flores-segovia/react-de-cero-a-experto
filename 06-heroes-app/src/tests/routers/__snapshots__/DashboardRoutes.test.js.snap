// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file 'DashboardRoutes.test.js' DashboardRoutes must be shown correctly 1`] = `
Array [
  <nav
    className="navbar navbar-expand-lg navbar-light bg-light"
  >
    <div
      className="container-fluid"
    >
      <a
        className="navbar-brand ms-2"
        href="/"
      >
        <img
          alt="heroes"
          className="me-2"
          height="30"
          src="/assets/icons/icons8-hulk-48.png"
          width="30"
        />
        Heroes
      </a>
      <button
        aria-controls="navbarText"
        aria-expanded="false"
        aria-label="Toggle navigation"
        className="navbar-toggler"
        data-bs-target="#navbarText"
        data-bs-toggle="collapse"
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="collapse navbar-collapse"
        id="navbarText"
      >
        <ul
          className="navbar-nav me-auto mb-2 mb-lg-0"
        >
          <a
            aria-current="page"
            className="nav-link active"
            href="/marvel"
            onClick={[Function]}
            style={Object {}}
          >
            Marvel Comics
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/dc"
            onClick={[Function]}
          >
            DC Comics
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/search"
            onClick={[Function]}
          >
            Search
          </a>
        </ul>
      </div>
      <div
        className="collapse navbar-collapse justify-content-end"
        id="navbarText"
      >
        <ul
          className="navbar-nav mb-2 mb-lg-0 align-items-center"
        >
          <span
            className="nav-item text-info me-3"
          >
            You're logged as 
            Jes√∫s
          </span>
          <button
            className="btn btn-danger"
            onClick={[Function]}
            style={
              Object {
                "color": "white",
                "textDecoration": "none",
              }
            }
            type="button"
          >
            Logout
          </button>
        </ul>
      </div>
    </div>
  </nav>,
  <div
    className="container mt-5"
  >
    <div>
      <h1>
        MarvelScreen
      </h1>
      <hr />
      <div
        className="row animate__animated animate__fadeIn"
      >
        <div
          className="col-xl-3 col-lg-3 col-md-4 col-sm-6 mb-4"
        >
          <div
            className="card h-100"
          >
            <img
              alt="Spider Man"
              className="card-img-top"
              src="./assets/heroes/marvel-spider.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Spider Man
              </h5>
              <p
                className="card-text"
              >
                Peter Parker
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Amazing Fantasy #15
                </small>
              </p>
              <a
                href="/hero/marvel-spider"
                onClick={[Function]}
              >
                See more
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-lg-3 col-md-4 col-sm-6 mb-4"
        >
          <div
            className="card h-100"
          >
            <img
              alt="Captain America"
              className="card-img-top"
              src="./assets/heroes/marvel-captain.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Captain America
              </h5>
              <p
                className="card-text"
              >
                Steve Rogers
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Captain America Comics #1
                </small>
              </p>
              <a
                href="/hero/marvel-captain"
                onClick={[Function]}
              >
                See more
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-lg-3 col-md-4 col-sm-6 mb-4"
        >
          <div
            className="card h-100"
          >
            <img
              alt="Iron Man"
              className="card-img-top"
              src="./assets/heroes/marvel-iron.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Iron Man
              </h5>
              <p
                className="card-text"
              >
                Tony Stark
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Tales of Suspense #39
                </small>
              </p>
              <a
                href="/hero/marvel-iron"
                onClick={[Function]}
              >
                See more
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-lg-3 col-md-4 col-sm-6 mb-4"
        >
          <div
            className="card h-100"
          >
            <img
              alt="Thor"
              className="card-img-top"
              src="./assets/heroes/marvel-thor.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Thor
              </h5>
              <p
                className="card-text"
              >
                Thor Odinson
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Journey into Myster #83
                </small>
              </p>
              <a
                href="/hero/marvel-thor"
                onClick={[Function]}
              >
                See more
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-lg-3 col-md-4 col-sm-6 mb-4"
        >
          <div
            className="card h-100"
          >
            <img
              alt="Hulk"
              className="card-img-top"
              src="./assets/heroes/marvel-hulk.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Hulk
              </h5>
              <p
                className="card-text"
              >
                Bruce Banner
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  The Incredible Hulk #1
                </small>
              </p>
              <a
                href="/hero/marvel-hulk"
                onClick={[Function]}
              >
                See more
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-lg-3 col-md-4 col-sm-6 mb-4"
        >
          <div
            className="card h-100"
          >
            <img
              alt="Wolverine"
              className="card-img-top"
              src="./assets/heroes/marvel-wolverine.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Wolverine
              </h5>
              <p
                className="card-text"
              >
                James Howlett
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  The Incredible Hulk #180
                </small>
              </p>
              <a
                href="/hero/marvel-wolverine"
                onClick={[Function]}
              >
                See more
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-lg-3 col-md-4 col-sm-6 mb-4"
        >
          <div
            className="card h-100"
          >
            <img
              alt="Daredevil"
              className="card-img-top"
              src="./assets/heroes/marvel-daredevil.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Daredevil
              </h5>
              <p
                className="card-text"
              >
                Matthew Michael Murdock
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Daredevil #1
                </small>
              </p>
              <a
                href="/hero/marvel-daredevil"
                onClick={[Function]}
              >
                See more
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-lg-3 col-md-4 col-sm-6 mb-4"
        >
          <div
            className="card h-100"
          >
            <img
              alt="Hawkeye"
              className="card-img-top"
              src="./assets/heroes/marvel-hawkeye.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Hawkeye
              </h5>
              <p
                className="card-text"
              >
                Clinton Francis Barton
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Tales of Suspense #57
                </small>
              </p>
              <a
                href="/hero/marvel-hawkeye"
                onClick={[Function]}
              >
                See more
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-lg-3 col-md-4 col-sm-6 mb-4"
        >
          <div
            className="card h-100"
          >
            <img
              alt="Cyclops"
              className="card-img-top"
              src="./assets/heroes/marvel-cyclops.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Cyclops
              </h5>
              <p
                className="card-text"
              >
                Scott Summers
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  X-Men #1
                </small>
              </p>
              <a
                href="/hero/marvel-cyclops"
                onClick={[Function]}
              >
                See more
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-lg-3 col-md-4 col-sm-6 mb-4"
        >
          <div
            className="card h-100"
          >
            <img
              alt="Silver Surfer"
              className="card-img-top"
              src="./assets/heroes/marvel-silver.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Silver Surfer
              </h5>
              <p
                className="card-text"
              >
                Norrin Radd
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  The Fantastic Four #48
                </small>
              </p>
              <a
                href="/hero/marvel-silver"
                onClick={[Function]}
              >
                See more
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
